### PR TITLE
"Server side support"

### PR DESCRIPTION
**Pull request by jplatte**
Or, more accurately: allow receiving requests and sending responses again!

Left to do:

* [x] Write docs for the new `SendRecv` trait
* [x] Add an example to `SendRecv`s documentation
* [x] Update docs for `ruma_api!`
* [x] Possibly rename `SendRecv`
* [x] ~~Decide whether we want the user to write `generics_indices` in the array case~~
* [x] ~~Potentially change the generated docs for `Incoming*` types~~
* [x] Remove leftover mentions of `#[wrap_incoming_inner]`
* [x] Better tests
* [x] Rewrite `#[wrap_incoming]` syntax (see comment below)
* [x] ~~Add support for `#[wrap_incoming]` on non-body fields?~~ Can be done later.
* [x] Code review?

Thanks @florianjacob for mulling over the generics issue with me!

### CODE REVIEW

**Comment by jplatte** [comment]
New `#[wrap_incoming]` syntax, as dicussed on [#ruma:matrix.org](https://matrix.to/#/#ruma:matrix.org):

```
#[wrap_incoming(
    // optionally select the type to wrap, otherwise the
    // whole type of the field is wrapped
    [type]
    // optionally select a generic type to use a the wrapper,
    // otherwise the incoming variant of `type` is used
    [with <wrapper_type>]
)]
```

Example:

```rust
#[derive(SendRecv)]
struct Test {
    pub foo: Foo,
    #[wrap_incoming] // or #[wrap_incoming()]
    pub bar: Bar,
    #[wrap_incoming(with EventResult)]
    pub x: XEvent,
    #[wrap_incoming(YEvent with EventResult)]
    pub ys: Vec<YEvent>,
    #[wrap_incoming(Baz)]
    pub wat: Option<Vec<HashMap<String, Baz>>>,
}

// generated
struct IncomingTest {
    pub foo: Foo,
    pub bar: IncomingBar,
    pub x: EventResult<XEvent>,
    pub ys: Vec<EventResult<YEvent>>,
    pub wat: Option<Vec<HashMap<String, IncomingBaz>>>,
}
```

**Comment by lu-fennell** [comment]
I had a look at the PR. The code itself is to involved for me to give any useful
comment right now. 

I think I get the idea of Outgoing/Incoming and I it's useful and nice, but I
think I would have had trouble understanding the use-case and purpose of
`Outgoing`, `Incoming` and `wrap_incoming` just by reading the docs (i.e.,
without watching the recording and reading the tests). Some notes in particular:

- in `ruma-api/src/lib.rs`, Section "Fallible deserialization": it is not clear
  when one would want to "allow fallible deserialization" and that the
  `#[wrap_incoming]` attribute should be used on fields. I think it is better to
  explain that here, shortly, before deferring to `Outgoing`s docs.
  
- in `ruma-api/src/lib.rs`, doc of `Outgoing`: it is not clear why one would
  need or want a Deserialize implementation. The doc should mention something
  like that there should always be an Incoming version of each Outgoing type,
  and that in the simplest case, when the type implements Deserialize, it is the
  type itself.
  
- in `ruma-api-macrocs/src/lib.rs`, doc of "derive_send_recv": it took me a
  while to get that `#[wrap_incoming]` is used for the nested case where the
  field's type requires fallible deserialization. The base case is always one of
  the `#[wrap_incoming(... with ..)] cases, right? That would be helpful to have
  in the docs.
  
  (.. also `derive_send_recv` should probably be `derive_outgoing`?)

File: ruma-api-macros/src/derive_outgoing.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                return true
+            }
+            _ => {}
+        }
+    }
+
+    false
+}
+
+fn impl_send_recv_incoming_self(ident: Ident) -> TokenStream {
```
Comment (Nauxuron) [inline ruma-api-macros/src/derive_outgoing.rs]: Should be `impl_outgoing_incoming_self`

File: ruma-api-macros/src/api/request.rs [hunk 2]
```diff
@@ -51,7 +75,6 @@ impl Request {
     pub fn has_header_fields(&self) -> bool {
         self.fields.iter().any(|field| field.is_header())
     }
-
```
Comment (Nauxuron) [inline ruma-api-macros/src/api/request.rs]: Why is this line deleted?

File: ruma-api-macros/src/lib.rs [hunk 3]
```diff
@@ -15,16 +15,64 @@ use std::convert::TryFrom as _;
...
+/// struct IncomingMyType {
+///     pub foo: Foo,
+///     pub bar: IncomingBar,
+///     pub baz: Option<IncomingBaz>,
+///     pub x: EventResult<XEvent>,
+///     pub ys: Vec<EventResult<YEvent>>,
+/// }
+/// ```
+#[proc_macro_derive(Outgoing, attributes(wrap_incoming, incoming_no_deserialize))]
+pub fn derive_send_recv(input: TokenStream) -> TokenStream {
```
Comment (Nauxuron) [inline ruma-api-macros/src/lib.rs]: As @lu-fennell already said, this one should be `derive_outgoing`.

**Comment by Nauxuron** [comment]
One thing that caught my eye:
`fn to_tokens` in `impl ToTokens for Api` is getting too long (426 lines) in my opinion. It's hard to follow and it would help if you could split it up in separate functions that are responsible for parts of that 109 line `quote!` block at the end.

File: ruma-api-macros/src/lib.rs (hunk 3 above)
Comment (jplatte) [reply ruma-api-macros/src/lib.rs]: Huh, I thought I had done that already.

**Comment by jplatte** [comment]
Yes, I've been working on refactoring the quote code actually. Although I'm not sure whether that function is going to become smaller with that.

File: ruma-api-macros/src/api/request.rs (hunk 2 above)
Comment (jplatte) [reply ruma-api-macros/src/api/request.rs]: No reason, don't know how that happened. Thanks for noticing.

**Comment by jplatte** [comment]
> * in `ruma-api/src/lib.rs`, Section "Fallible deserialization": it is not clear
>   when one would want to "allow fallible deserialization" and that the
>   `#[wrap_incoming]` attribute should be used on fields. I think it is better to
>   explain that here, shortly, before deferring to `Outgoing`s docs.

I've added a TODO comment about this. Since ruma-api isn't really user-facing, I think it's fine to merge this without the documentation fully in place. I'll also open an issue about it.

> * in `ruma-api/src/lib.rs`, doc of `Outgoing`: it is not clear why one would
>   need or want a Deserialize implementation. The doc should mention something
>   like that there should always be an Incoming version of each Outgoing type,
>   and that in the simplest case, when the type implements Deserialize, it is the
>   type itself.

Same for this.

> * in `ruma-api-macrocs/src/lib.rs`, doc of "derive_send_recv": it took me a
>   while to get that `#[wrap_incoming]` is used for the nested case where the
>   field's type requires fallible deserialization. The base case is always one of
>   the `#[wrap_incoming(... with ..)] cases, right? That would be helpful to have
>   in the docs.

With base case, do you mean the innermost part of a request that uses `derive(Outgoing)` and `#[wrap_incoming]`? In that case yes, that's always one with using the "with" syntax (although not necessarily with anything before that keyword, see e.g. [here](https://github.com/ruma/ruma-client-api/pull/51/files#diff-bba4d53a50b216a33875175fe0bd0208L197-L207)). I've added another TODO comment for that.

> (.. also `derive_send_recv` should probably be `derive_outgoing`?)

Yes. Fixed.

**Comment by lu-fennell** [comment]
> With base case, do you mean the innermost part of a request that uses derive(Outgoing) and #[wrap_incoming]?

Yes, that's what I meant.
